{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68058e45",
   "metadata": {},
   "source": [
    "# pyPAGE example notebook: bulk and single-cell\n",
    "\n",
    "This notebook demonstrates two end-to-end workflows:\n",
    "\n",
    "1. **Bulk-like analysis** using bundled example differential expression and GO annotations.\n",
    "2. **Single-cell-like analysis** where differential scores are computed from two cell groups with a Mann-Whitney U test, then passed to `pyPAGE`.\n",
    "\n",
    "> Runtime note: parameters below are tuned for interactive speed. Increase `n_shuffle` for publication-scale analyses.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e6188069",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matplotlib is building the font cache; this may take a moment.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "from pypage import PAGE, ExpressionProfile, GeneSets\n",
    "\n",
    "np.random.seed(0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c338dfe2",
   "metadata": {},
   "source": [
    "## 1) Bulk example\n",
    "\n",
    "Load differential expression bins and GO gene sets from `example_data/`, then run `PAGE`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3a18b476-040d-4138-a599-7ec261ca0550",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AP2S1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CNP</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAVIN1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FUOM</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOL4L</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  bin\n",
       "0   AP2S1    0\n",
       "1     CNP    0\n",
       "2  CAVIN1    2\n",
       "3    FUOM    0\n",
       "4   NOL4L    0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df = pd.read_csv(\n",
    "    \"../example_data/AP2S1.tab.gz\",\n",
    "    sep=\"\t\",\n",
    "    header=None,\n",
    "    names=[\"gene\", \"bin\"],\n",
    ")\n",
    "expr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c067eeb7-c3d4-4a7c-bd08-cce069fa7e0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "bin\n",
       "1    17026\n",
       "2      283\n",
       "0      233\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_df.bin.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "72aee03d-fffb-4311-94c6-a6fcb846977b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Expression Profile\n",
       ">> num_genes: 17542\n",
       ">> num_bins: 10"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_bulk = ExpressionProfile(expr_df[\"gene\"].values, expr_df[\"bin\"].values, is_bin=True)\n",
    "exp_bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3e644549-ad91-4b76-9b01-3756fa779c07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SDF2L1</td>\n",
       "      <td>'de novo' posttranslational protein folding (G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HSPA9</td>\n",
       "      <td>'de novo' posttranslational protein folding (G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCT2</td>\n",
       "      <td>'de novo' posttranslational protein folding (G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ST13</td>\n",
       "      <td>'de novo' posttranslational protein folding (G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HSPA6</td>\n",
       "      <td>'de novo' posttranslational protein folding (G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene                                            pathway\n",
       "0  SDF2L1  'de novo' posttranslational protein folding (G...\n",
       "1   HSPA9  'de novo' posttranslational protein folding (G...\n",
       "2    CCT2  'de novo' posttranslational protein folding (G...\n",
       "3    ST13  'de novo' posttranslational protein folding (G...\n",
       "4   HSPA6  'de novo' posttranslational protein folding (G..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df = pd.read_csv(\n",
    "    \"../example_data/GO_BP_2021_index.txt.gz\",\n",
    "    sep=\"\t\",\n",
    "    header=None,\n",
    "    names=[\"gene\", \"pathway\"],\n",
    ")\n",
    "go_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "80ae7fb5-39b6-4489-b268-73da58daa5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "ont_bulk = GeneSets(go_df[\"gene\"].values, go_df[\"pathway\"].values, n_bins=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "482793b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.\n",
      "calculating conditional mutual information: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2272/2272 [00:05<00:00, 421.93it/s]\n",
      "permutation testing:   9%|█████████████▍                                                                                                                                        | 203/2272 [00:12<02:10, 15.89it/s]\n",
      "consolidating redundant pathways: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2272/2272 [00:05<00:00, 396.12it/s]\n",
      "hypergeometric tests: 57it [00:00, 1066.83it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bulk pathways returned: 57\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>CMI</th>\n",
       "      <th>p-value</th>\n",
       "      <th>Regulation pattern</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mitotic spindle organization (GO:0007052)</td>\n",
       "      <td>0.004761</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rRNA metabolic process (GO:0016072)</td>\n",
       "      <td>0.004278</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mitotic sister chromatid segregation (GO:0000070)</td>\n",
       "      <td>0.003973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>generation of neurons (GO:0048699)</td>\n",
       "      <td>0.003268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>positive regulation of DNA biosynthetic proces...</td>\n",
       "      <td>0.003160</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>positive regulation of protein localization to...</td>\n",
       "      <td>0.003078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>protein stabilization (GO:0050821)</td>\n",
       "      <td>0.002708</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>axonogenesis (GO:0007409)</td>\n",
       "      <td>0.002569</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>positive regulation of cell junction assembly ...</td>\n",
       "      <td>0.002509</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>anterior/posterior pattern specification (GO:0...</td>\n",
       "      <td>0.002219</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             pathway       CMI  p-value  \\\n",
       "0          mitotic spindle organization (GO:0007052)  0.004761      0.0   \n",
       "1                rRNA metabolic process (GO:0016072)  0.004278      0.0   \n",
       "2  mitotic sister chromatid segregation (GO:0000070)  0.003973      0.0   \n",
       "3                 generation of neurons (GO:0048699)  0.003268      0.0   \n",
       "4  positive regulation of DNA biosynthetic proces...  0.003160      0.0   \n",
       "5  positive regulation of protein localization to...  0.003078      0.0   \n",
       "6                 protein stabilization (GO:0050821)  0.002708      0.0   \n",
       "7                          axonogenesis (GO:0007409)  0.002569      0.0   \n",
       "8  positive regulation of cell junction assembly ...  0.002509      0.0   \n",
       "9  anterior/posterior pattern specification (GO:0...  0.002219      0.0   \n",
       "\n",
       "   Regulation pattern  \n",
       "0                   1  \n",
       "1                   1  \n",
       "2                   1  \n",
       "3                  -1  \n",
       "4                   1  \n",
       "5                   1  \n",
       "6                   1  \n",
       "7                  -1  \n",
       "8                  -1  \n",
       "9                  -1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Optional speed filter for notebook use\n",
    "ont_bulk.filter_pathways(min_size=20, max_size=400)\n",
    "\n",
    "p_bulk = PAGE(\n",
    "    exp_bulk,\n",
    "    ont_bulk,\n",
    "    n_shuffle=50,\n",
    "    alpha=0.01,\n",
    "    k=7,\n",
    "    filter_redundant=True,\n",
    "    n_jobs=1,\n",
    ")\n",
    "\n",
    "bulk_results, bulk_hm = p_bulk.run()\n",
    "print(f\"Bulk pathways returned: {len(bulk_results)}\")\n",
    "bulk_results.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1bbc02f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize top pathways\n",
    "if bulk_hm is not None:\n",
    "    bulk_hm.show(max_rows=30, title=\"Bulk example\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e83bc30",
   "metadata": {},
   "source": [
    "## 2) Single-cell example\n",
    "\n",
    "Create two synthetic cell populations (`A` vs `B`), compute per-gene differential scores with a Mann-Whitney U test, and run `PAGE` on those scores.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4b16fba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a gene universe from ontology genes and sample a subset for speed\n",
    "all_ont_genes = pd.unique(go_df[\"gene\"])\n",
    "n_genes = 2500\n",
    "genes_sc = np.random.choice(all_ont_genes, size=n_genes, replace=False)\n",
    "\n",
    "n_cells_a = 60\n",
    "n_cells_b = 60\n",
    "\n",
    "# Baseline expression per gene\n",
    "baseline = np.random.gamma(shape=2.0, scale=1.0, size=n_genes)\n",
    "\n",
    "# Simulate two groups with targeted perturbations\n",
    "cells_a = np.random.poisson(lam=np.clip(baseline, 0.05, None), size=(n_cells_a, n_genes))\n",
    "perturbed = baseline.copy()\n",
    "perturbed[:120] *= 1.8   # up in B\n",
    "perturbed[120:240] *= 0.5  # down in B\n",
    "cells_b = np.random.poisson(lam=np.clip(perturbed, 0.05, None), size=(n_cells_b, n_genes))\n",
    "\n",
    "\n",
    "def differential_score_u_test(group_a: np.ndarray, group_b: np.ndarray) -> np.ndarray:\n",
    "    # Compute sign(mean diff) * (1 - pvalue) per gene.\n",
    "    n = group_a.shape[1]\n",
    "    score = np.zeros(n, dtype=float)\n",
    "    sign = np.sign(group_b.mean(axis=0) - group_a.mean(axis=0))\n",
    "    for i in range(n):\n",
    "        if np.array_equal(np.unique(group_a[:, i]), np.unique(group_b[:, i])):\n",
    "            score[i] = 0.0\n",
    "        else:\n",
    "            p = mannwhitneyu(group_a[:, i], group_b[:, i], alternative=\"two-sided\").pvalue\n",
    "            score[i] = sign[i] * (1.0 - p)\n",
    "    return score\n",
    "\n",
    "scores_sc = differential_score_u_test(cells_a, cells_b)\n",
    "\n",
    "exp_sc = ExpressionProfile(genes_sc, scores_sc, n_bins=10)\n",
    "\n",
    "# Restrict annotation to selected genes for this toy single-cell run\n",
    "go_sc = go_df[go_df[\"gene\"].isin(set(genes_sc))].copy()\n",
    "ont_sc = GeneSets(go_sc[\"gene\"].values, go_sc[\"pathway\"].values, n_bins=6)\n",
    "ont_sc.filter_pathways(min_size=15, max_size=300)\n",
    "\n",
    "p_sc = PAGE(\n",
    "    exp_sc,\n",
    "    ont_sc,\n",
    "    n_shuffle=30,\n",
    "    alpha=0.01,\n",
    "    k=7,\n",
    "    filter_redundant=True,\n",
    "    n_jobs=1,\n",
    ")\n",
    "\n",
    "sc_results, sc_hm = p_sc.run()\n",
    "print(f\"Single-cell pathways returned: {len(sc_results)}\")\n",
    "sc_results.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "748f98c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "if sc_hm is not None:\n",
    "    sc_hm.show(max_rows=30, title=\"Single-cell example\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcd7fd93",
   "metadata": {},
   "source": [
    "## Next steps\n",
    "\n",
    "- Increase `n_shuffle` (for example, 500 to 1000+) for stronger p-value estimates.\n",
    "- Replace synthetic single-cell matrices with your real cell-by-gene matrix and group labels.\n",
    "- Optionally convert accessions with `convert_from_to(...)` if expression and annotations use different gene ID types.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
